WordPress/gutenberg
# Try/even better previews

This PR has been cloned from https://github.com/WordPress/gutenberg/pull/16113 in order to achieve the same goal but from a different approach.

In short, the final idea behind of these PRs is improving the visualization of the `<BlockPreview />`.

### Some screenshots

**<Button />**

<img src="https://user-images.githubusercontent.com/77539/62347778-3745da00-b4d1-11e9-8e3c-c6804036f6f8.png" width="500px" />

**Preview styles in the sidebar**
<img src="https://user-images.githubusercontent.com/77539/62347672-d74f3380-b4d0-11e9-9174-332c839f576b.png" width="500px" />

**Full preview**
<img src="https://user-images.githubusercontent.com/77539/62347725-01a0f100-b4d1-11e9-9e35-8c28b4e78d4e.png" width="500px" />

## How can use it

Unlike the previous approach, this proposal does not require that the dimensions of the block (`srcWidth` and `srcHeight`), the scale factor be defined (`scaleFactor`), and whether or not the scale should be adjusted (`fitToScale`). So, we only need to pass the `blocks` property to the `<BlockPreview />` component to make it work.

```jsx
<BlockPreviewContent blocks={ cloneBlock( block, { className: styleClassName } ) } />
</div>
```

## How does it work

### Width and the Scale.

The `scale` is computed **only** considering the width of the blocks and its container. We could polish this approach incorporating the possibility to fit the preview in height, centered, etc. I image something like the `background-size` and `background-position`.

The scaled is computed dividing the width of the container element (it belongs to the `<PreviewBlock />` component) by the width of the block to visualize. **VERY IMPORTANT: it works with only one block so far**. However, I'm quite confident that we can make it work for more many blocks as well.

```
const scale = containerWidth / blocksWidth;
```

To get the width of the container we use the `useRef()` hook and the `ref` property. Kind simple.

Getting the width of the block to preview is kinda hacky at the moment. There is something funny when we try to get the DOM element. Although we are using the `useLayoutEffect()` hook, the dom isn't there yet meaning that when we try to pick it up through its clientId (using a dom/util function) it returns null.
First hack: setTimeout(). Wrapping the code into this function we can get the DOM element. I'd like to think that we can get a better solution, digging the reason why it happens. It seems to be related to the `<BlockEditorProvider />` block. I've followed advice from @youknowriad, creating a new component to wrap the `<BlockList />` but didn't work either. Maybe I'm missing something.

Another relevant change that I'm introducing here is using a CSS class to identify what part(elements) of the Block we want to visualize. AFAIK, This is completely needed since we don't have any other way to detect the exact part of the component that we want.

Although the `getBlockDOMNode()` returns the DOM Element of the block container, generally it doesn't match which what we'd like to visualize. And we can't generalize among all blocks saying ... _just pick the first child of its children_, or something like that.
Adding this new `.wp-block-preview-container` to those elements of the blocks, we can run a query selection easily. For this, I've added the `getClockPreviewContainerDOMNode()` function.
It means that we need to explicitly add this class the blocks that can be previewed. This PR adds the class to the `table`, `button` and `pullquote` component. While this seems to be complicated, it could also be considered one that allows us to easily define which elements we want to visualize from the block.

Finally, once we have the container width and the block with, the rest is a matter of simple calculation.

# How to test

* create a post with a `table`, `pullquote` and/or a `button`.
* check how looks the preview in the different locations: Sidebar, Switcher and the Full Preview Popover.

# Status
Pretty sure there are a lot of things to improve/change and at the end, I'm not 100% confident about this approach neither.

![g-preview](https://user-images.githubusercontent.com/77539/62349835-492a7b80-b4d7-11e9-99f1-e3316ba44a66.gif)
